### PR TITLE
PMM-7779 remove old database server versions

### PR DESCRIPTION
**Pull request by jprukner**
[PMM-7779](https://jira.percona.com/browse/PMM-7779)

It's based on #706.

- [ ] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1636

- [ ] Links to other linked pull requests (optional).
---
- [ ] Tests passed.
- [ ] Feature build pass.
- [ ] (Re)requested review.
- [ ] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

### CODE REVIEW

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Looks like you pulled outdated version of components service

File: services/management/dbaas/components_service.go [hunk 1]
```diff
@@ -96,6 +106,71 @@ func (c componentsService) GetPXCComponents(ctx context.Context, req *dbaasv1bet
...
+	switch operator {
+	case PSMDBOperator:
+		versions = m.Mongod
+	case PXCOperator:
+		versions = m.Pxc
+	default:
+		return errors.Errorf("failed to filter out disabled versions, unknown operator %s", operator)
+	}
+
+	versionsToDelete := make([]string, 0)
```
Comment (Dasio) [inline services/management/dbaas/components_service.go:158]:
```suggestion
	var versionsToDelete []string
```

File: services/management/dbaas/components_service.go [hunk 2]
```diff
@@ -96,6 +106,71 @@ func (c componentsService) GetPXCComponents(ctx context.Context, req *dbaasv1bet
...
+	disabledVersionsPerOperator := disabledVersions[operator]
+	for version := range versions {
+		filterOut, err := shouldFilterOutVersion(disabledVersionsPerOperator, version)
+		if err != nil {
+			return err
+		}
+		if filterOut {
+			versionsToDelete = append(versionsToDelete, version)
```
Comment (Dasio) [inline services/management/dbaas/components_service.go:166]: do we need this? you can remove it from map here

**Comment by jprukner** [comment]
@BupycHuk Rebased on latest

File: services/management/dbaas/components_service.go [hunk 3]
```diff
@@ -97,6 +107,71 @@ func (c componentsService) GetPXCComponents(ctx context.Context, req *dbaasv1bet
...
+func (c componentsService) removeDisabledVersions(operator string, m matrix) error {
+	// TODO: load this map from database.
+	disabledVersions := map[string](map[string]disablingModifier){
+		PXCOperator: map[string]disablingModifier{
+			"5.7": equal,
+		},
+		PSMDBOperator: map[string]disablingModifier{
+			"4.2": less,
+		},
+	}
```
Comment (Dasio) [inline services/management/dbaas/components_service.go:147]:
```suggestion
	disabledVersions := map[string]map[string]disablingModifier{
		PXCOperator: {
			"5.7": equal,
		},
		PSMDBOperator: {
			"4.2": less,
		},
	}
```

**Review by Dasio — APPROVED** [review]

File: services/management/dbaas/components_service.go (hunk 3 above)
Comment (jprukner) [reply services/management/dbaas/components_service.go]: done

File: services/management/dbaas/components_service.go (hunk 2 above)
Comment (jprukner) [reply services/management/dbaas/components_service.go]: removed

File: services/management/dbaas/components_service.go (hunk 1 above)
Comment (jprukner) [reply services/management/dbaas/components_service.go]: don't need this, removed the var

File: services/management/dbaas/components_service.go [hunk 4]
```diff
@@ -34,6 +37,13 @@ type componentsService struct {
 	versionServiceClient versionService
 }
 
+type disablingModifier int
+
+const (
+	equal disablingModifier = iota
```
Comment (BupycHuk) [inline services/management/dbaas/components_service.go:43]: We don't need this enum, we can just remove everything less then 8.0.0 for PXC

File: services/management/dbaas/components_service.go [hunk 5]
```diff
@@ -105,6 +176,10 @@ func (c componentsService) versions(ctx context.Context, params componentsParams
 
 	versions := make([]*dbaasv1beta1.Version, 0, len(components.Versions))
 	for _, v := range components.Versions {
+		err = c.removeDisabledVersions(params.operator, v.Matrix)
```
Comment (BupycHuk) [inline services/management/dbaas/components_service.go:179]: My suggestion would be to just pass minimum version into `c.matrix` and take care it there.
For example: `c.matrix(v.Matrix.Mongod, "4.2.0")`

File: services/management/dbaas/components_service.go [hunk 6]
```diff
@@ -24,6 +24,9 @@ import (
 	"github.com/sirupsen/logrus"
 	"gopkg.in/reform.v1"
 
+	goversion "github.com/hashicorp/go-version"
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply services/management/dbaas/components_service.go:27]: let's use your library in both places, because `pmm` comparer has some issues.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/management/dbaas/components_service.go (hunk 5 above)
Comment (jprukner) [reply services/management/dbaas/components_service.go:179]: We should be able to filter out single version.

Comment (BupycHuk) [reply services/management/dbaas/components_service.go:179]: Not in the scope of this ticket

File: services/management/dbaas/components_service.go (hunk 4 above)
Comment (jprukner) [reply services/management/dbaas/components_service.go:39]: removed

File: services/management/dbaas/components_service.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jprukner) [reply services/management/dbaas/components_service.go]: Import moved to the block above.

I replaced the pmm version with hashicorp one.

File: services/management/dbaas/components_service.go (hunk 5 above)
Comment (jprukner) [reply services/management/dbaas/components_service.go]: Cool, at least It will be simple.

File: services/management/dbaas/components_service.go [hunk 7]
```diff
@@ -125,23 +126,26 @@ func (c componentsService) versions(ctx context.Context, params componentsParams
 	return versions, nil
 }
 
-func (c componentsService) matrix(m map[string]component) map[string]*dbaasv1beta1.Component {
+func (c componentsService) matrix(m map[string]component, minimalVersion ...*goversion.Version) map[string]*dbaasv1beta1.Component {
```
Comment (BupycHuk) [inline services/management/dbaas/components_service.go:129]: I would suggest to do this
```suggestion
func (c componentsService) matrix(m map[string]component, minimalVersion *goversion.Version) map[string]*dbaasv1beta1.Component {
```

and below just check if it's nil or not

Comment (jprukner) [reply services/management/dbaas/components_service.go]: Right, I changed it.

File: services/management/dbaas/components_service.go [hunk 8]
```diff
@@ -19,12 +19,11 @@ package dbaas
 import (
 	"context"
 
+	goversion "github.com/hashicorp/go-version"
+	"github.com/percona/pmm-managed/models"
```
Comment (BupycHuk) [inline services/management/dbaas/components_service.go:23]: to the third block

Comment (jprukner) [reply services/management/dbaas/components_service.go]: moved

**Review by BupycHuk — APPROVED** [review]

**Review by palash25 — APPROVED** [review]

**Comment by jprukner** [comment]
rebased on PMM-2.0

File: services/management/dbaas/components_service.go [hunk 9]
```diff
@@ -125,23 +127,26 @@ func (c componentsService) versions(ctx context.Context, params componentsParams
...
+func (c componentsService) matrix(m map[string]component, minimalVersion *goversion.Version) map[string]*dbaasv1beta1.Component {
 	result := make(map[string]*dbaasv1beta1.Component)
 
 	var lastVersion string
-	lastVersionParsed := version.MustParse("0.0.0")
+	lastVersionParsed, _ := goversion.NewVersion("0.0.0")
 	for v, component := range m {
+		parsedVersion, err := goversion.NewVersion(v)
+		if err != nil {
+			c.l.Warnf("couldn't parse version %s: %s", v, err.Error())
```
Comment (jprukner) [inline services/management/dbaas/components_service.go:138]: we should continue to next version here

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**